### PR TITLE
Adding role checks, services, API clients and config to wire up the create-and-vary-a-licence-api, community-api and prison-api

### PR DESCRIPTION
**Pull request by timharrison-moj**
* Adds API clients for community-api, prison-api and create-and-vary-a-licence-api.
* Adds basic templates to view data returned from APIs.
* Adds simple routes and services to test out getting/showing data from API clients.
* Adds unit tests for routes and API clients.
* Adds configuration for 3 x API clients.
* Adds role-checks for access to the service - ROLE_CVL_OMU and/or ROLE_GLOBAL_SEARCH (for now).
* Add Helm configuration for API client URLs,
* Most of this stuff will evolve for real user needs - currently just implementing the basic framework with trial forms/routes.

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -50,11 +50,10 @@ services:
       - REDIS_HOST=redis
       - HMPPS_AUTH_EXTERNAL_URL=http://localhost:9090/auth
       - HMPPS_AUTH_URL=http://hmpps-auth:8080/auth
-      # These will need to match new creds in the seed auth service auth
       - API_CLIENT_ID=create-and-vary-a-licence-client
-      - API_CLIENT_SECRET=client-secret
+      - API_CLIENT_SECRET=clientsecret
       - SYSTEM_CLIENT_ID=create-and-vary-a-licence-admin
-      - SYSTEM_CLIENT_SECRET=system-client-secret
+      - SYSTEM_CLIENT_SECRET=client_secret
```
Comment (timharrison-moj) [inline docker-compose.yml:57]: Just updated to match the local values you added in hmpps-auth seed data.

File: .eslintrc.json [hunk 2]
```diff
@@ -33,6 +33,8 @@
         "plugin:prettier/recommended"
       ],
       "rules": {
+        "no-shadow": "off",
+        "@typescript-eslint/no-shadow": "error",
         "@typescript-eslint/no-use-before-define": 0,
```
Comment (timharrison-moj) [inline .eslintrc.json:38]: Required for the way in which services are imported as a list.

File: server/app.ts [hunk 3]
```diff
@@ -31,10 +32,14 @@ export default function createApp(userService: UserService): express.Application
...
 
-  app.use('/', indexRoutes(standardRouter(userService)))
+  // CSRF protection
+  if (!testMode) {
+    app.use(csurf())
+  }
 
+  app.use(authorisationMiddleware)
+  app.use(setupRoutes(services))
   app.use((req, res, next) => next(createError(404, 'Not found')))
```
Comment (timharrison-moj) [inline server/app.ts:43]: Replaces all application route setup with a single call to `setupRoutes(services)`

File: server/data/communityApiClient.ts [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  async getStaffDetailByUsername(username: string): Promise<StaffDetail> {
+    return this.restClient.get({ path: `/secure/staff/username/${username}` }) as Promise<StaffDetail>
+  }
+
+  async getStaffCaseload(staffId: number): Promise<ManagedOffender[]> {
+    return this.restClient.get({
+      path: `/secure/staff/staffIdentifier/${staffId}/managedOffenders`,
+    }) as Promise<ManagedOffender[]>
+  }
+}
```
Comment (timharrison-moj) [inline server/data/communityApiClient.ts:21]: Just some trial endpoints for now - I will probably need to get the caseload, and staff details.

File: server/data/communityClientTypes.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  currentPom: boolean
+  currentRo: boolean
+  nomsNumber: string
+  offenderId: number
+  offenderSurname: string
+  omEndDate: string
+  omStartDate: string
+  staffCode: string
+  staffIdentifier: number
+}
```
Comment (timharrison-moj) [inline server/data/communityClientTypes.ts:46]: This is taken from the UOF method of defining the API responses as interfaces.

File: server/index.ts [hunk 6]
```diff
@@ -1,11 +1,6 @@
 import createApp from './app'
-import HmppsAuthClient from './data/hmppsAuthClient'
-import TokenStore from './data/tokenStore'
-import UserService from './services/userService'
+import { services } from './services'
```
Comment (timharrison-moj) [inline server/index.ts:3]: Services are all created in `services/index.ts`

File: server/middleware/authorisationMiddleware.ts [hunk 7]
```diff
@@ -1,22 +1,33 @@
...
 import logger from '../../logger'
 
-export default function authorisationMiddleware(authorisedRoles: string[] = []): RequestHandler {
-  return (req, res, next) => {
-    if (res.locals && res.locals.user && res.locals.user.token) {
-      const { authorities: roles = [] } = jwtDecode(res.locals.user.token) as { authorities?: string[] }
+export enum AuthRole {
+  OMU = 'ROLE_CVL_OMU',
+  GLOBAL_SEARCH = 'ROLE_GLOBAL_SEARCH',
+}
```
Comment (timharrison-moj) [inline server/middleware/authorisationMiddleware.ts:8]: Defines an enum of all the the valid roles to allow into the service.

File: server/routes/index.ts [hunk 8]
```diff
@@ -1,13 +1,42 @@
...
+    get('/', (req, res) => {
+      res.render('pages/index')
+    })
+
+  const otherRoutes = () => {
+    get('/test/data', otherAccessRoutes.listTestData)
+    get('/staff/:username/detail', otherAccessRoutes.getStaffDetail)
+    get('/staff/:staffId/caseload', otherAccessRoutes.getStaffCaseload)
+    get('/prisoner/:nomsId/detail', otherAccessRoutes.getPrisonerDetail)
+  }
```
Comment (timharrison-moj) [inline server/routes/index.ts:27]: This is the pin-phone method, which I liked, of defining routes as `functions of related routes`, and then defining these in their own modules. The `OtherRoutes` is a temporary split - so I'll probably split them based on the user roles - e.g. probation officer routes, prison approver routes, offender management unit routes etc..

File: server/routes/testutils/appSetup.ts [hunk 9]
```diff
@@ -31,30 +32,50 @@ class MockUserService extends UserService {
...
+    userService: new MockUserService(),
+    prisonerService: {} as PrisonerService,
+    licenceService: {} as LicenceService,
+    communityService: {} as CommunityService,
+    ...overrides,
+  })
   auth.default.authenticationMiddleware = () => (req, res, next) => next()
-  return appSetup(allRoutes(standardRouter(new MockUserService())), production)
+  return appSetup({ router, userRoles, production })
 }
```
Comment (timharrison-moj) [inline server/routes/testutils/appSetup.ts:81]: I'm not entirely happy with this `appSetup()` for unit testing yet. But it works for now, and I'll continually improve as and when I see issues.

File: server/views/pages/managedOffenders.njk [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+                      <td class="govuk-table__cell">{{ offender.nomsNumber }}</td>
+                      <td class="govuk-table__cell">{{ offender.offenderSurname }}</td>
+                    </tr>
+                  {%  endfor %}
+                </tbody>
+
+            </table>
+        </div>
+    </div>
+{% endblock %}
```
Comment (timharrison-moj) [inline server/views/pages/managedOffenders.njk:44]: A temporary form to show API client results

File: server/views/pages/prisonerDetail.njk [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+                    <td class="govuk-table__cell">{{ prisonerDetail.latestLocationId }}</td>
+                    <td class="govuk-table__cell">{{ prisonerDetail.pncNumber }}</td>
+                    <td class="govuk-table__cell">{{ prisonerDetail.croNumber }}</td>
+                  </tr>
+                </tbody>
+
+            </table>
+        </div>
+    </div>
+{% endblock %}
```
Comment (timharrison-moj) [inline server/views/pages/prisonerDetail.njk:40]: A temporary form to show API client results

File: server/views/pages/staffDetail.njk [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+                    <td class="govuk-table__cell">{{ staffDetail.email }}</td>
+                    <td class="govuk-table__cell">{{ staffDetail.username }}</td>
+                    <td class="govuk-table__cell">{{ staffDetail.telephoneNumber }}</td>
+                  </tr>
+                </tbody>
+
+            </table>
+        </div>
+    </div>
+{% endblock %}
```
Comment (timharrison-moj) [inline server/views/pages/staffDetail.njk:43]: A temporary form to show API client results

File: server/views/pages/testData.njk [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+                    <td class="govuk-table__cell">{{ test.key }}</td>
+                    <td class="govuk-table__cell">{{ test.value }}</td>
+                  </tr>
+                {% endfor %}
+              </tbody>
+
+            </table>
+        </div>
+    </div>
+{% endblock %}
```
Comment (timharrison-moj) [inline server/views/pages/testData.njk:33]: A temporary form to show API client results

**Review by joelstobart-moj — APPROVED** [review]
LGTM

**Review by yasinmustafa — APPROVED** [review]
LGTM

**Pull request merged**